### ISSUE TITLE
Setting percentile of best trials for importance calculations

### ISSUE THREAD

**Issue by oasidorshin** [optuna/optuna#4985]
### Motivation

When looking at parameters importances plots, a lot of the times the most important parameters will be the ones with badly selected search spaces, e.g.:

Weight decay search space in [1e-06, 1], and values closer to 1 are leading to blatant underfit and low accuracy. Despite usually low impact of this parameter, because of bad values and their huge impact on accuracy weight decay will be in the top of param importances plot.

This is of cource misleading and only showing trivial dependencies. A lot of the times you want to ignore bad trials and only look at the best ones to see which parameters contribute more to actually good function values

### Description

My proposed solution is to add top_percentile parameter to `optuna.visualization.plot_param_importances`, so that only top percentile of trials will be present during importances calculations. Also, for better interpretability, we can even set the default value to something around 50%, although of course this is a matter for discussion

### Additional context (optional)

I can work on this issue if the idea is approved

**Comment by contramundum53** [comment]
Thanks for your suggestion!

The current algorithms underlying parameter importance (fANOVA, Mean Decrease Impurity, SHAP) are inherently unstable unless the parameters are sampled uniformly (i.e. with `RandomSampler`, `QMCSampler` or `BruteForceSampler`). Thus, for real-world use cases, importance calculation is subject to many heuristics.

I think it is worth considering an interface for users to specify which trials to use, as this is more general. @nabenabe0928 @HideakiImamura What do you think?

**Comment by HideakiImamura** [comment]
Thank you for your suggestion. As Wang has mentioned, I believe many users would be pleased if such a function is supported. On the other hand, I think there is room for discussion regarding the design. Should we allow users to specify the top percentile of objective function values as you have suggested, or should we let them specify the threshold value of the objective function, or should we provide function to filter the study itself...? There are various options, so allow us to discuss the appropriate implementation among committers.

Lastly, let me introduce a simple workaround. For example, if you only want to consider trials where the objective function value is less than or equal to `x`, you can use `add_trials` as follows:

```python
study = optuna.create_study()
study.optimize(objective, n_trials=100)

filtered_study = optuna.create_study()
filtered_trials = [t for t in study.trials if t.value <= x]
filtered_study.add_trials(filtered_trials)

optuna.visualization.plot_param_importances(filtered_study)
```

**Comment by oasidorshin** [comment]
@contramundum53 Yes, I understand that parameter importance algorithms are inherently biased unless the parameters are sampled uniformly, but I think that calculating the importance while discarding worst trials is useful in many real world cases, because often we want to understand which parameters are important for getting the best results, not for filtering out really bad cases.

@HideakiImamura I think that percentile is the most universal and interpretable for general usage - so that users can easily specify it once for their different runs and not tune individually for each function/run/project. It also allows us to set some sensible default value (e.g. top 75% trials)

Thanks for the workaround! I think that this feature in general would be quite easy to implement :)

**Comment by nabenabe0928** [comment]
Sorry for the late response, I completely missed the thread.
It is already kinda late, but I add some comments from my side in terms of the quantification itself:
So as mentioned earlier, f-ANOVA is biased when the distribution of an objective function value heavily leans on the worse side and it is very effective to remove such trials from the f-ANOVA computation.

The easiest way to achieve it is to clip the function values like `min(threshold, trial.value)` [1,2] (originally proposed by Hutter et al. in Section 3.3):

```python
import optuna
import numpy as np


def objective(trial):
    x = trial.suggest_float("x", -5, 5)
    y = trial.suggest_float("y", -5, 5)
    if x ** 2 + y ** 2 >= 3:
        # Both are equally important in the global space.
        return x ** 2 + y ** 2
    else:
        # `x` is more important in the top-performing space.
        return x ** 2 + 1e-2 * y ** 2


def modify_trial(trial, threshold):
    return optuna.trial.create_trial(
        params=trial.params,
        distributions=trial.distributions,
        # Thresholding to reduce the effect from bad spaces.
        value=min(threshold, trial.value),
    )


def plot_importance(quantile):
    n_trials = 100
    study = optuna.create_study()
    study.optimize(objective, n_trials=n_trials)
    threshold = np.sort([t.value for t in study.trials])[int(quantile * n_trials) - 1]

    study_for_importance = optuna.create_study()
    modified_trials = [
        modify_trial(t, threshold)
        for t in study.trials
    ]
    study_for_importance.add_trials(modified_trials)
    fig = optuna.visualization.plot_param_importances(study_for_importance)
    fig.show()

```

The example above is taken from Eq. (31) by Watanabe et al.[2] and the function is designed so that both parameters show equal importance in the global space while `x` becomes more important in top-performing spaces.

The result of `plot_importance` with `quantile=1.0` (the default Optuna implementation):

![newplot (2)](https://github.com/optuna/optuna/assets/47781922/d21fa5e1-655b-42bf-a7e0-3340f1c30697)

And here is after the modification with `quantile=0.1` and `filter=True`:

![newplot (3)](https://github.com/optuna/optuna/assets/47781922/5910a2b8-4be5-4adc-8f64-611b531525ab)

Lastly, the result with `quantile=0.1` and `filter=False`:

![newplot (4)](https://github.com/optuna/optuna/assets/47781922/d8bdf94f-850e-4085-b1d9-b79bb213136e)

HOWEVER, I must note that this quantification is still biased as discussed by Watanabe et al. [2], so please notice that when you use it.
The bias reduction [2] can be easily implemented as can be seen [here](https://github.com/nabenabe0928/local-anova) if you are interested.

**References**
[1] [An Efficient Approach for Assessing Hyperparameter Importance](https://proceedings.mlr.press/v32/hutter14.html)
[2] [PED-ANOVA: Efficiently Quantifying Hyperparameter Importance in Arbitrary Subspaces](https://arxiv.org/abs/2304.10255)
